### PR TITLE
Implements tests for 4.2.2-A,B and C

### PR DESCRIPTION
**Pull request by dbernstein**
4.2.2-A,B and C
    4.2.3, 4.2.4, and 4.2.5

### CODE REVIEW

File: src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java [hunk 1]
```diff
@@ -88,33 +109,41 @@ public void ldprmOptionsMaySupportDelete(final String uri) {
...
                                         ps);
-
         //create an LDPRm
+        final String mementoURI = createMemento(uri, info);
+        final Response mementoResponse = doGet(mementoURI);
         //verify that a subsequent OPTIONS does not return POST
+        confirmAbsenceOfHeaderValueInMultiValueHeader("Allow", "POST", mementoResponse);
         //verify that issuing a POST returns a 405 (method not allowed).
+        final Response response = doPostUnverified(mementoURI);
+        assertEquals(response.getStatusCode(), 405, "Response to a POST should be 405");
```
Comment (bseeger) [inline src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java]: The 405 seems to be implementation specific  - perhaps it's what's people would use, but I'm not sure if it should be assumed.  Thoughts?

Comment (bseeger) [reply src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java]: It's not specified in the spec, so perhaps this should just check for a 4XX error?

File: src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java [hunk 2]
```diff
@@ -123,16 +152,22 @@ public void ldprmMustNotSupportPut(final String uri) {
                                         ps);
 
         //create an LDPRm
+        final String mementoURI = createMemento(uri, info);
+        final Response mementoResponse = doGet(mementoURI);
         //verify that a subsequent OPTIONS does not return PUT
+        confirmAbsenceOfHeaderValueInMultiValueHeader("Allow", "PUT", mementoResponse);
         //verify that issuing a PUT returns a 405 (method not allowed).
+        final Response response = doPutUnverified(mementoURI);
+        assertEquals(response.getStatusCode(), 405, "Response to a PUT should be 405");
```
Comment (bseeger) [inline src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java]: same comment as above about the specific 405 error - should probably just check for a 4xx error.

File: src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java [hunk 3]
```diff
@@ -141,8 +176,14 @@ public void ldprmMustNotSupportPatch(final String uri) {
                                         ps);
 
         //create an LDPRm
+        final String mementoURI = createMemento(uri, info);
+        final Response mementoResponse = doGet(mementoURI);
         //verify that a subsequent OPTIONS does not return PATCH
+        confirmAbsenceOfHeaderValueInMultiValueHeader("Allow", "PATCH", mementoResponse);
         //verify that issuing a PATCH returns a 405 (method not allowed).
+        final Response response = doPatchUnverified(mementoURI);
+        assertEquals(response.getStatusCode(), 405, "Response to a PATCH should be 405");
```
Comment (bseeger) [inline src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java]: same comment about the 405 error - should probably just check for a 4xx error.

File: src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java [hunk 4]
```diff
@@ -52,16 +58,27 @@ public void ldprmMustSupportOptions(final String uri) {
                                         ps);
 
         //create an LDPRm
+        final String mementoURI = createMemento(uri, info);
+        final Response mementoResponse = doGet(mementoURI);
         //verify that a subsequent GET returns OPTIONS in "Allow" header.
+        confirmPresenceOfHeaderValueInMultiValueHeader("Allow", "OPTIONS", mementoResponse);
+    }
 
+    private String createMemento(final String uri, final TestInfo info) {
```
Comment (bseeger) [inline src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java]: I think something like this should go into the AbstractVersioningTest.java file.  But, I don't think it should do things - namely create a versioned resource and create a memento.  I'd say pass in the URI of the versioned resource, and only have this create the memento.  

I'd like to use this function in my PR, hence wanting it up in the Abstract Versioning class.  It's pretty handy.

File: src/main/java/org/fcrepo/spec/testsuite/versioning/AbstractVersioningTest.java [hunk 5]
```diff
@@ -61,12 +61,19 @@ protected Response createVersionedResource(final String uri, final TestInfo info
     }
 
     protected void confirmPresenceOfHeaderValueInMultiValueHeader(final String headerName, final String headerValue,
-                                                                  final Response response) {
+                                                                       final Response response) {
```
Comment (bseeger) [inline src/main/java/org/fcrepo/spec/testsuite/versioning/AbstractVersioningTest.java]: should revert this change.

File: src/main/java/org/fcrepo/spec/testsuite/versioning/AbstractVersioningTest.java [hunk 6]
```diff
@@ -61,12 +61,19 @@ protected Response createVersionedResource(final String uri, final TestInfo info
...
         Assert
             .assertTrue(hasHeaderValueInMultiValueHeader(headerName, headerValue, response),
                         headerName + " with a value of " + headerValue + " must be present but is not!");
     }
 
+    protected void confirmAbsenceOfHeaderValueInMultiValueHeader(final String headerName, final String headerValue,
+                                                                  final Response response) {
```
Comment (bseeger) [inline src/main/java/org/fcrepo/spec/testsuite/versioning/AbstractVersioningTest.java]: picky, I know, but there's an extra space on this line before 'final'  :)

File: src/main/java/org/fcrepo/spec/testsuite/versioning/AbstractVersioningTest.java (hunk 5 above)
Comment (dbernstein) [reply src/main/java/org/fcrepo/spec/testsuite/versioning/AbstractVersioningTest.java]: @bseeger :  good catch.

File: src/main/java/org/fcrepo/spec/testsuite/versioning/AbstractVersioningTest.java (hunk 6 above)
Comment (dbernstein) [reply src/main/java/org/fcrepo/spec/testsuite/versioning/AbstractVersioningTest.java]: picky is good.

Comment (dbernstein) [reply src/main/java/org/fcrepo/spec/testsuite/versioning/AbstractVersioningTest.java]: picky is good.

File: src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java (hunk 4 above)
Comment (dbernstein) [reply src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java]: @bseeger 👍

File: src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java (hunk 1 above)
Comment (dbernstein) [reply src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java]: @bseeger : I'm not sure either.   It seems to me that anything other than 405 would indicate some other problem.  I would suggest that we keep it as 405 and change it if we run into complaints from one or more Fedora implementers.

**Comment by dbernstein** [comment]
@bseeger : feel free to merged it when you're ready.

File: src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java (hunk 1 above)
Comment (awoods) [reply src/main/java/org/fcrepo/spec/testsuite/versioning/LdprmHttpOptions.java]: I agree that 405 is the appropriate response here. 
As a point of reference for other tests that need to support 4xx status codes, there is this helper:
https://github.com/fcrepo/Fedora-API-Test-Suite/blob/master/src/main/java/org/fcrepo/spec/testsuite/AbstractTest.java#L360
